### PR TITLE
MediaLibrary: directory types

### PR DESCRIPTION
**Pull request by kurzee**
This first PR is the base for the subsequent Media exporter PRs, which will be merged here before merging into `develop`.

In this PR we're introducing the idea of using specific types of local Media directories. Between `.uploads`, `.cache`, and `.temporary` we're able to control where/how Media files are stored on disk and the inherent handling of the files based on iOS.

We'll be using `.uploads` to hang on to Media files that need to be uploaded. The `.cache` is planned to be used for downloading existing Media files that can be discarded by iOS or the user. And lastly, `.temporary` is handy for not worrying too much about cleaning up anything written to disk, as seen with unit tests for example.

Currently the default is `.uploads` which uses the same directory path we've always used.

Note: I included a goofy default `func` as `localDirectory()` and `makeLocalMediaURL()` to satiate Obj-C.

To test:
1. Build, make sure everything builds.
2. Run all unit tests, make sure everything passes.
3. Test that inserting and posting an image from the device library picker works.
4. Test that inserting and posting an image from the Media library picker works.
5. Test that the internal Media Library collection loads as expected.

Needs review: @frosty and @SergioEstevao ready for round one 😅

### CODE REVIEW

File: WordPress/Classes/Extensions/MediaLibrary+Local.swift [hunk 1]
```diff
@@ -6,34 +6,53 @@ extension MediaLibrary {
...
+        if fileManager.fileExists(atPath: mediaDirectory.path, isDirectory: &isDirectory) == false || isDirectory.boolValue == false {
+            try fileManager.createDirectory(at: mediaDirectory, withIntermediateDirectories: true, attributes: nil)
         }
-        return media
+        return mediaDirectory
+    }
+
+    /// Helper method for ObjC as a default method with type.
+    ///
+    class func localDirectory() throws -> URL {
```
Comment (frosty) [inline WordPress/Classes/Extensions/MediaLibrary+Local.swift]: Perhaps rename this to `localUploadsDirectory()`?

File: WordPress/Classes/Extensions/MediaLibrary+Local.swift [hunk 2]
```diff
@@ -47,6 +66,12 @@ extension MediaLibrary {
         return url.incrementalFilename()
     }
 
+    /// Helper method for ObjC as a default method with type.
+    ///
+    class func makeLocalMediaURL(withFilename filename: String, fileExtension: String?) throws -> URL {
```
Comment (frosty) [inline WordPress/Classes/Extensions/MediaLibrary+Local.swift]: Perhaps rename this to `makeLocalMediaUploadsURL` or similar?

File: WordPress/Classes/Extensions/MediaLibrary+Local.swift [hunk 3]
```diff
@@ -143,7 +168,7 @@ extension MediaLibrary {
                         filesToKeep.insert(localThumbnailURL.lastPathComponent)
                     }
                 }
-                try purgeLocalDirectory(exceptFiles: filesToKeep)
+                try purgeLocalDirectory(exceptFiles: filesToKeep, type: .uploads)
```
Comment (frosty) [inline WordPress/Classes/Extensions/MediaLibrary+Local.swift]: Do we want to clean out caches / tmp too?

File: WordPress/WordPressTest/MediaLibraryTests.swift [hunk 4]
```diff
@@ -6,12 +6,55 @@ class MediaLibraryTests: XCTestCase {
...
+            XCTFail("Error accessing or creating local media cache directory: \(error)")
+        }
+    }
+
+    func testThatLocalMediaDirectoryAsTemporaryIsAvailable() {
+        do {
+            let url = try MediaLibrary.localDirectory(.temporary)
+            assertThatMediaDirectoryIsNamedMedia(url: url)
+            let fileManager = FileManager.default
+            // The .cache directory should be within the system Caches directory.
```
Comment (frosty) [inline WordPress/WordPressTest/MediaLibraryTests.swift]: Copy paste from caches test

File: WordPress/Classes/Extensions/MediaLibrary+Local.swift [hunk 5]
```diff
@@ -6,34 +6,53 @@ extension MediaLibrary {
...
         // If the filepath does not exist, or if the filepath does exist but it is not a directory, try creating the directory.
         // Note: This way, if unexpectedly a file exists but it is not a dir, an error will throw when trying to create the dir.
         var isDirectory: ObjCBool = false
-        if fileManager.fileExists(atPath: media.path, isDirectory: &isDirectory) == false || isDirectory.boolValue == false {
-            try fileManager.createDirectory(at: media, withIntermediateDirectories: true, attributes: nil)
-            var resourceValues = URLResourceValues()
-            resourceValues.isExcludedFromBackup = true
-            try media.setResourceValues(resourceValues)
+        if fileManager.fileExists(atPath: mediaDirectory.path, isDirectory: &isDirectory) == false || isDirectory.boolValue == false {
+            try fileManager.createDirectory(at: mediaDirectory, withIntermediateDirectories: true, attributes: nil)
```
Comment (SergioEstevao) [inline WordPress/Classes/Extensions/MediaLibrary+Local.swift]: With attributes at nil, is the folder excluded or included on backups?

File: WordPress/Classes/Extensions/MediaLibrary+Local.swift (hunk 3 above)
Comment (kurzee) [inline WordPress/Classes/Extensions/MediaLibrary+Local.swift]: Yep! Haven't implemented the different cleanup methods yet, based on the the directory types.

File: WordPress/Classes/Extensions/MediaLibrary+Local.swift (hunk 5 above)
Comment (kurzee) [inline WordPress/Classes/Extensions/MediaLibrary+Local.swift]: With this implementation, both the caches and tmp directories won't back up, or iOS will disregard the contents as it sees fit.

For the uploads directory, I avoided explicitly disabling backups as we technically want to hang on to these contents for any pending uploads.

If the user were to have a couple of local posts with Media ready for upload, but restored their phone from a backup for whatever reason, we'll still have the pending files. That was the idea anyways, do you think we should go ahead and exclude from backup regardless? @SergioEstevao

Comment (frosty) [inline WordPress/Classes/Extensions/MediaLibrary+Local.swift]: I think it makes sense to allow those pending uploads to be backed up 👍

Comment (SergioEstevao) [inline WordPress/Classes/Extensions/MediaLibrary+Local.swift]: It makes sense.

**Comment by kurzee** [comment]
Ready for another look 🚀

**Comment by kurzee** [comment]
Merging everything into `develop`, thanks so much for diving into this @frosty and @SergioEstevao!

**Pull request merged**